File: 2024/14/calibrate.rb
#!/usr/bin/env ruby

# 2024 Day 7 Part 1 / Puzzle 2024/13

class Puzzle
  def initialize()
    # each entry is an array of [total, [ inputs ]]
    @calibrations = [ ]
  end

  def decode(line)
    total, numstr = line.split(/: /, 2)
    nums = numstr.split(' ').map(&:to_i)
    @calibrations.push([total.to_i, nums])
  end

  def calibrations
    @calibrations
  end
end

def check_total(val, tot, remainders)
  if val > tot
    return false
  end
  if remainders.empty?
    return val == tot
  end

  entry, rest = remainders[0], remainders[1..]
  res = (check_total(entry + val, tot, rest) or
         check_total(entry * val, tot, rest) or
         check_total((val.to_s + entry.to_s).to_i, tot, rest)
        )
  return res
end

p = Puzzle.new()

File.foreach(ARGV.shift || '../13/input') { |line|
  p.decode(line.chomp)
}

sum = p.calibrations.filter_map { |entry|
  total, nums = entry

  total if check_total(0, total, nums)
}.sum

puts "#{sum}"

File: 2024/14/calibrate.rs
use std::env;
use std::fs;

// Advent of Code: 2024, day 7, part 2

fn check_total(val: u64, tot: u64, remainders: &[u64]) -> bool {
    if val > tot {
        return false;
    }
    if remainders.is_empty() {
        return val == tot;
    }

    if let Some((entry, rest)) = remainders.split_first() {
        let concat_val = (val.to_string() + &entry.to_string()).parse::<u64>().unwrap();
        return
            check_total(val + entry, tot, rest) ||
            check_total(val * entry, tot, rest) ||
            check_total(concat_val, tot, rest)
            ;
    }
    else {
        panic!("This should be impossible");
    }
}

fn test_line(line: &String) -> u64 {
    let tot_and_nums = line.split(": ").collect::<Vec<_>>();
    let total = tot_and_nums[0].parse::<u64>().unwrap();
    let nums = tot_and_nums[1]
        .split(' ')
        .map(str::parse::<u64>)
        .map(Result::unwrap)
        .collect::<Vec<_>>();

    if check_total(0, total, &nums) {
        return total;
    }

    return 0;
}

fn main() {
    let default_filename: &'static str = "../13/input";
    let args: Vec<String> = env::args().collect();
    let in_file: String = match args.len() {
        // No args provided save argv[0]
        1 => String::from(default_filename),
        _ => args[1].clone(),
    };

    let lines = fs::read_to_string(&in_file)
        .expect("Should have been able to read the file")
        .lines()
        .map(String::from)
        .collect::<Vec<_>>();

    let sum: u64 = lines.iter().map(test_line).sum();

    println!("{}", sum);
}
